docs/src/submodules/Utilities/reference.md

```diff
@@ -220,4 +220,6 @@ Utilities.DoubleDicts.DoubleDict
 Utilities.DoubleDicts.DoubleDictInner
 Utilities.DoubleDicts.IndexDoubleDict
 Utilities.DoubleDicts.IndexDoubleDictInner
+Utilities.DoubleDicts.outer_keys
+Utilities.DoubleDicts.outer_keys_vector
```

Comment:
I guess one question I had: we we need both of these? What about making the second a keyword argument to `outer_keys`?

Reply:
This works just as well, and if you want the vector you can call `collect`?
````Julia
"""
    outer_keys(d::AbstractDoubleDict; only_nonempty::Bool = false)

Return an iterator over the outer keys of the `AbstractDoubleDict` `d`.

Each outer key is a `Tuple{Type,Type}` so that a double loop can
be easily used (with a function-barrier to ensure the inner loop is
type-stable):
```julia
function function_barrier(dict, ::Type{F}, ::Type{S}) where {F,S}
    for (k, v) in dict[F, S]
        # ...
    end
end
```

If `only_nonempty = true`, only outer keys that have a non-empty set of inner
keys will be returned.
"""
function outer_keys(d::AbstractDoubleDict; only_nonempty::Bool = false)
    return ((F, S) for ((F, S), d) in d.dict if !only_nonempty || length(d) > 0)
end
````

Reply:
This is not Lazy.

I like both versions because one is Lazy and the other is very useful for things like ListOfConstraintTypePresent.

One way to simplify would be:

Keep the current `outer_keys` and add:
```julia
function nonempty_outer_keys(d::AbstractDoubleDict)
    return ((F, S) for ((F, S), d) in d.dict if length(d) > 0)
end
```